### ISSUE TITLE
Should validation tests care about layout order?

### ISSUE THREAD

**Issue by mknyszek** [domokit/mojo#813]
Since decoding Mojoms just involves using pointers, why do we care about the layout? Pointers should always be non-negative, but other than that why is it an error to have an earlier pointer point to something later? The same thing goes for handles: why does there need to be an order? Specifically, this issue is in reference to the validation tests with wrong_layout_order and wrong_handle_order in the name.

The Rust bindings right now can handle out-of-order layouts without issue. In fact, in the Rust bindings I actually rely on out-of-order-layout behavior in order to efficiently encoding Rust maps. The issue is I need to consume the map on encode (in order to preserve move semantics for handles) and so its more efficient for me to encode keys and values in an interleaved manner as there is no easy way to extract associative arrays out of a Rust hashmap without additional allocations (these allocations are otherwise wasteful since the buffers would get consumed immediately anyway).

**Comment by j9brown** [comment]
Ordering is a simple way to prevent aliasing.  If we allow aliasing to
occur, we could break assumptions made by recipients of data structures as
they attempt to mutate them.

**Comment by vardhan-cr** [comment]
Yeah, objects are encoded in depth-first order of when they are referenced, and I think that decision was made for a few reasons, given that we don't want two pointers pointing to the same object (although having this is beneficial, it comes with some caveats like what @j9brown mentioned above).

The big reason I think is: It makes validation easier and faster:

Since objects are in DFS and only have one pointer to them, you can validate regions of the wire format after you've consumed them by simply moving a "claimed memory" pointer forward. Otherwise, you have to maintain slices of "claimed" memory regions, and make sure none of them overlap as you claim further regions. Since validation is something you do on all incoming messages (and will be done on all outgoing message), making the objects be in DFS-order makes validation memory-bounded and much faster.

I think the big advantage of objects being encoded in a non-DFS order is if the programmer wants to construct objects in whatever order, and they don't want any copying done when encoding.  But you can say that it's the minority case (and really only possible in C/C++), and make validation faster instead.
